File: app/src/main/java/com/pawels96/skyrimperkcalculator/ui/BuildAdapter.java
package com.pawels96.skyrimperkcalculator.ui;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.text.SpannableString;
import android.text.SpannableStringBuilder;
import android.text.style.ForegroundColorSpan;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.PopupMenu;
import android.widget.TextView;

import com.pawels96.skyrimperkcalculator.R;
import com.pawels96.skyrimperkcalculator.enums.SkillType;
import com.pawels96.skyrimperkcalculator.models.Build;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static com.pawels96.skyrimperkcalculator.Utils.PREFS_MULTIPLIER;
import static com.pawels96.skyrimperkcalculator.Utils.PREFS_NAME;

public class BuildAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<Build> items;
    private Context context;

    public BuildAdapter(List<Build> items, Context context) {
        this.items = items;
        this.context = context;

        stealth = context.getResources().getColor(R.color.skillStealthBright);
        combat = context.getResources().getColor(R.color.skillCombatBright);
        magic = context.getResources().getColor(R.color.skillMagicBright);
        multiplier = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).getFloat(PREFS_MULTIPLIER, 1f);
    }

    public Build getItem(int index){
        return items.get(index);
    }

    private int stealth, combat, magic;
    private float multiplier;
    private BuildAdapterCallback callback;

    private String currentBuildName;

    public void setCurrentBuildName(String currentBuildName) {
        this.currentBuildName = currentBuildName;
    }

    public String getCurrentBuildName() {
        return currentBuildName;
    }

    public interface BuildAdapterCallback {
        void onRename(Build build);

        void onDelete(Build build, boolean canDelete);

        void showDescription(Build build);

        void onClick(Build build);
    }

    public void setCallback(BuildAdapterCallback callback) {
        this.callback = callback;
    }

    public void delete(Build build) {
        notifyItemRemoved(items.indexOf(build));
        items.remove(build);
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new Holder(LayoutInflater.from(viewGroup.getContext()).inflate(i, viewGroup, false));
    }

    @Override
    public int getItemViewType(int position) {
        return R.layout.list_item_build;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int i) {

        final Build build = items.get(i);
        final Holder holder = (Holder) viewHolder;

        holder.root.setOnClickListener(v -> callback.onClick(build));

        if (build.getName().equals(currentBuildName)) {
            SpannableString str0 = new SpannableString(build.getName());
            str0.setSpan(new ForegroundColorSpan(magic), 0, str0.length(), 0);
            holder.name.setText(str0);
        } else holder.name.setText(build.getName());

        String lvl = Integer.toString(build.getRequiredLevel(multiplier));

        int magicPerks = build.getPerkDistribution().get(SkillType.MAGIC);
        int stealthPerks = build.getPerkDistribution().get(SkillType.STEALTH);
        int combatPerks = build.getPerkDistribution().get(SkillType.COMBAT);

        SpannableStringBuilder builder = new SpannableStringBuilder(context.getString(R.string.perks) + ": ");

        SpannableString str1 = new SpannableString(combatPerks + " ");
        str1.setSpan(new ForegroundColorSpan(combat), 0, str1.length(), 0);
        builder.append(str1);

        SpannableString str2 = new SpannableString(magicPerks + " ");
        str2.setSpan(new ForegroundColorSpan(magic), 0, str2.length(), 0);
        builder.append(str2);

        SpannableString str3 = new SpannableString(Integer.toString(stealthPerks));
        str3.setSpan(new ForegroundColorSpan(stealth), 0, str3.length(), 0);
        builder.append(str3);

        holder.perks.setText(builder, TextView.BufferType.SPANNABLE);
        String levelText = context.getString(R.string.level) + ": " + lvl;
        holder.level.setText(levelText);

        String description = build.getDescription();

        if (description == null || description.isEmpty())
            holder.description.setVisibility(View.GONE);
        else {
            holder.description.setVisibility(View.VISIBLE);

            holder.description.setText(description);
        }

        holder.button.setOnClickListener(v -> {

            PopupMenu popup = new PopupMenu(context, holder.button);

            popup.getMenuInflater().inflate(R.menu.menu_list_item, popup.getMenu());

            if (items.size() < 2)
                popup.getMenu().findItem(R.id.delete).setVisible(false);

            popup.setOnMenuItemClickListener(item -> {

                int menuId = item.getItemId();

                switch (menuId) {
                    case R.id.rename:
                        callback.onRename(build);
                        break;

                    case R.id.delete:
                        callback.onDelete(build, items.size() != 1);
                        break;

                    case R.id.description:
                        callback.showDescription(build);
                        break;
                }
                return true;
            });
            popup.show();
        });

    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    private static class Holder extends RecyclerView.ViewHolder {

        public Holder(View view) {
            super(view);

            root = view;
            name = view.findViewById(R.id.build_name);
            button = view.findViewById(R.id.context_menu);
            level = view.findViewById(R.id.level);
            perks = view.findViewById(R.id.build_info);
            description = view.findViewById(R.id.description);
        }

        View root;
        Button button;
        TextView name;
        TextView level;
        TextView perks;
        TextView description;
    }
}

File: app/src/main/java/com/pawels96/skyrimperkcalculator/presentation/BuildAdapter.kt
package com.pawels96.skyrimperkcalculator.presentation

import android.content.Context
import android.text.SpannableStringBuilder
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.PopupMenu
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.pawels96.skyrimperkcalculator.R
import com.pawels96.skyrimperkcalculator.domain.Build
import com.pawels96.skyrimperkcalculator.domain.SkillType

class BuildAdapter(private val context: Context, private val callback: BuildAdapterCallback) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val items = mutableListOf<Build>()
    private var selected: Int = 0

    private val stealth = context.resources.getColor(R.color.skillStealthBright)
    private val combat = context.resources.getColor(R.color.skillCombatBright)
    private val magic = context.resources.getColor(R.color.skillMagicBright)
    private val vampire = context.resources.getColor(R.color.skillVampireBright)
    private val werewolf = context.resources.getColor(R.color.skillWerewolfBright)

    private var multiplier = 0f

    interface BuildAdapterCallback {
        fun onRename(build: Build)
        fun onDelete(build: Build)
        fun showDescription(build: Build)
        fun onClick(build: Build)
    }

    fun display(items: List<Build>, selectedIndex: Int, multiplier: Float) {
        this.items.clear()
        this.items.addAll(items)
        this.multiplier = multiplier
        selected = selectedIndex
        notifyDataSetChanged()
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, i: Int) {
        val build = items[i]
        val holder = viewHolder as Holder

        holder.root.setOnClickListener { v: View? -> callback.onClick(build) }
        holder.name.text = if (selected == i) build.name.colored(magic) else build.name

        val skills = mutableMapOf(
                stealth to build.getPerkDistribution()[SkillType.STEALTH],
                combat to build.getPerkDistribution()[SkillType.COMBAT],
                magic to build.getPerkDistribution()[SkillType.MAGIC],
                vampire to build.vampireSkill[build.vampirePerkSystem]?.selectedPerksCount,
                werewolf to build.werewolfSkill[build.werewolfPerkSystem]?.selectedPerksCount
        )
                .filter { it.value!! > 0 }
                .map { "${it.value} ".colored(it.key) }

        val builder = SpannableStringBuilder(context.getString(R.string.perks) + ": ")

        if (skills.isNotEmpty()) {
            skills.forEach { builder.append(it) }
        }
        else builder.append("0")

        holder.perks.text = builder//, TextView.BufferType.SPANNABLE)

        val lvl = build.getRequiredLevel(multiplier).toString()
        val levelText = context.getString(R.string.level).toString() + ": " + lvl
        holder.level.text = levelText

        val description = build.description
        if (description.isEmpty()) holder.description.visibility = View.GONE else {
            holder.description.visibility = View.VISIBLE
            holder.description.text = description
        }

        holder.button.setOnClickListener { v: View? ->
            val popup = PopupMenu(context, holder.button)
            popup.menuInflater.inflate(R.menu.menu_list_item, popup.menu)

            if (items.size < 2) popup.menu.findItem(R.id.delete).isVisible = false
            popup.setOnMenuItemClickListener { item: MenuItem ->
                when (item.itemId) {
                    R.id.rename -> callback.onRename(build)
                    R.id.delete -> callback.onDelete(build)
                    R.id.description -> callback.showDescription(build)
                }
                true
            }
            popup.show()
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return Holder(LayoutInflater.from(parent.context).inflate(viewType, parent, false))
    }

    override fun getItemViewType(position: Int): Int = R.layout.list_item_build

    override fun getItemCount(): Int = items.size

    private class Holder(var root: View) : RecyclerView.ViewHolder(root) {
        var button: Button = root.findViewById(R.id.context_menu)
        var name: TextView = root.findViewById(R.id.name_edit)
        var level: TextView = root.findViewById(R.id.level)
        var perks: TextView = root.findViewById(R.id.build_info)
        var description: TextView = root.findViewById(R.id.description)
    }
}